### PR TITLE
Update vagrant to include kubeworkers and refator edge, worker loop

### PR DESCRIPTION
**Pull request by andreijs**
Hey guys,

Here is the updated vagrant file for kubeworker.

- [done ] Installs cleanly on a fresh build of most recent master branch
- [ done] Upgrades cleanly from the most recent release
- [ done] Updates documentation relevant to the changes

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -26,6 +30,36 @@ else
  config_hash = config_hash.merge(YAML.load(File.read(config_path)))
 end
 
+def spin_up(config_hash:, config:, server_array:, hostvars:, hosts:, server_type:)
```
Comment (siddharthist) [inline Vagrantfile]: This function could use an explaining comment

**Comment by siddharthist** [comment]
This looks great, thanks @andreijs! I left just one comment.

File: Vagrantfile [hunk 2]
```diff
@@ -26,6 +30,36 @@ else
...
+
+      hosts += "#{ip}    #{hostname}\n"
+      # Update Ansible variables
+      server_array << hostname
+      server_hostvars = {
+        hostname => {
+          "ansible_ssh_host" => ip,
+          "private_ipv4" => ip,
+          "public_ipv4" => ip,
+          "role" => server_type
```
Comment (siddharthist) [inline Vagrantfile]: server_type could just be renamed to "role" for consistency

File: Vagrantfile [hunk 3]
```diff
@@ -44,59 +78,9 @@ Vagrant.configure(2) do |config|
...
-          "private_ipv4" => ip,
-          "public_ipv4" => ip,
-          "role" => "edge"
-        }
-      }
-      hostvars.merge!(edge_hostvars)
-    end
-  end
+  spin_up(config_hash: config_hash, config: config, hostvars: hostvars, server_array: workers, hosts: hosts, server_type: 'worker')
+  spin_up(config_hash: config_hash, config: config, hostvars: hostvars, server_array: workers, hosts: hosts, server_type: 'kubeworker')
```
Comment (siddharthist) [inline Vagrantfile]: Shouldn't this `server_array` be `kubeworkers` or similar?

**Comment by siddharthist** [comment]
We should also probably add a line for kubeworkers here: 
https://github.com/CiscoCloud/mantl/pull/1365/files#diff-23b6f443c01ea2efcb4f36eedfea9089L141

**Comment by andreimc** [comment]
@siddharthist thanks for your feedback I had made the changes pointed out 👍

File: roles/consul/defaults/main.yml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ---
-consul_package: consul-0.6.3
-consul_ui_package: consul-ui-0.6.3
+consul_package: consul-0.6.4
+consul_ui_package: consul-ui-0.6.4
```
Comment (siddharthist) [inline roles/consul/defaults/main.yml]: Can we put this in a separate PR? Doesn't seem directly related.

File: vagrant/README.rst [hunk 5]
```diff
@@ -24,23 +24,23 @@ Variables
...
-.. data:: worker_ip_start, control_ip_start, edge_ip_start
+.. data:: worker_ip_start, control_ip_start, edge_ip_start, kubeworker_ip_start
 
    A base IP address which will have its last digit appended. For example, if
    ``worker_ip_start`` is set to "192.168.100.10", the first worker node will
    have the IP address 192.168.100.101, the second will have 192.168.100.102,
    etc.
 
-.. data:: worker_memory, control_memory, edge_memory
+.. data:: worker_memory, control_memory, edge_memory, kubeworker_memeory
```
Comment (ryane) [inline vagrant/README.rst]: small typo: s/memeory/memory

**Comment by siddharthist** [comment]
I got this error on `vagrant up`:
```
TASK: [kubernetes-addons | create/update skydns replication controller] ******* 
failed: [control-01] => {"failed": true}
msg: error running kubectl (/bin/kubectl --server=http://localhost:8085/ --namespace=kube-system create --filename=/etc/kubernetes/manifests/skydns-rc.yaml) command (rc=1): Error from server: error when creating "/etc/kubernetes/manifests/skydns-rc.yaml": namespaces "kube-system" not found


FATAL: all hosts have already failed -- aborting
```

File: roles/consul/defaults/main.yml (hunk 4 above)
Comment (andreimc) [inline roles/consul/defaults/main.yml]: Sorry I didn't realize I commited this

**Comment by andreimc** [comment]
I get this one some times haven't seen the one you got @siddharthist 

```
TASK: [kubernetes-addons | create/update elasticsearch service] ***************
failed: [control-01] => {"failed": true}
msg: error running kubectl (/bin/kubectl --server=http://localhost:8085/ --namespace=kube-system create --filename=/etc/kubernetes/manifests/es-svc.yaml) command (rc=1): Error from server: error when creating "/etc/kubernetes/manifests/es-svc.yaml": Internal error occurred: failed to allocate a serviceIP: cannot allocate resources of type serviceipallocations at this time

```

**Comment by andreimc** [comment]
Everything works for me apart from Kube UI and nginx-consul starting on kubeworker ref #1346

```
PLAY RECAP ********************************************************************
mesos | wait for zookeeper service to be registered -------------------- 38.28s
common | install system utilities -------------------------------------- 37.09s
consul | wait for leader ----------------------------------------------- 30.54s
kubernetes | pull hyperkube docker image ------------------------------- 24.81s
kubernetes-master | wait for apiserver to come up ---------------------- 16.20s
etcd | restart skydns -------------------------------------------------- 14.56s
kubernetes-master | download kubernetes binaries ----------------------- 14.36s
mantlui | ensure nginx-mantlui docker image is present ----------------- 12.17s
kubernetes-node | download kubernetes binaries ------------------------- 12.07s
zookeeper | install zookeepercli package ------------------------------- 10.47s
control-01                 : ok=271  changed=206  unreachable=0    failed=0
edge-001                   : ok=122  changed=91   unreachable=0    failed=0
kubeworker-001             : ok=152  changed=106  unreachable=0    failed=0
localhost                  : ok=0    changed=0    unreachable=0    failed=0
worker-001                 : ok=123  changed=88   unreachable=0    failed=0

➜  mantl (master) ✔ vagrant ssh kubeworker-001
No vagrant-config.yml found, using defaults
Last login: Wed Apr 20 23:54:41 2016 from control-01
[vagrant@kubeworker-001 ~]$
```

**Comment by siddharthist** [comment]
I got another error this time:
```
TASK: [kubernetes-addons | create/update grafana service] ********************* 
failed: [control-01] => {"failed": true}
msg: error running kubectl (/bin/kubectl --server=http://localhost:8085/ --namespace=kube-system create --filename=/etc/kubernetes/manifests/grafana-service.yaml) command (rc=1): Error from server: error when creating "/etc/kubernetes/manifests/grafana-service.yaml": Internal error occurred: failed to allocate a serviceIP: cannot allocate resources of type serviceipallocations at this time


FATAL: all hosts have already failed -- aborting
```

**Comment by andreimc** [comment]
@siddharthist can you give me your machine specs OS, ansible version etc ?

**Comment by siddharthist** [comment]
@andreimc I have Vagrant 1.8.1 and Oracle VM VirtualBox Manager 5.0.16_OSE. My host's version of ansible doesn't/shouldn't affect anything, the VMs are provisioned from the control node.

**Comment by andreimc** [comment]
@siddharthist I really don't know why it fails, me and a co-worker both tried to spin it up and it worked fine, Vagrant file updates should not really cause ansible to fail ... maybe get someone else to try it.

**Comment by ryane** [comment]
I'm also having a lot of trouble getting kubernetes to run on Vagrant. Provisioning fails intermittently with various errors. Here are a couple I have seen repeatedly:

```
TASK: [kubernetes-addons | create or update dashboard] ************************
failed: [control-01] => {"failed": true}
msg: error running kubectl (/bin/kubectl --server=http://localhost:8085/ --namespace=kube-system create --filename=/etc/kubernetes/manifests/kubernetes-dashboard.yaml) command (rc=1): You have exposed your service on an external port on all nodes in your
cluster.  If you want to expose this service to the external internet, you may
need to set up firewall rules for the service port(s) (tcp:30000) to serve traffic.

See http://releases.k8s.io/release-1.2/docs/user-guide/services-firewalls.md for more details.
service "kubernetes-dashboard" created
```

```
TASK: [kubernetes-addons | create or update dashboard] ************************
failed: [control-01] => {"failed": true}
msg: error running kubectl (/bin/kubectl --server=http://localhost:8085/ --namespace=kube-system create --filename=/etc/kubernetes/manifests/kubernetes-dashboard.yaml) command (rc=1): replicationcontroller "kubernetes-dashboard" created
```

Repeated provisioning attempts might ultimately complete but still seeing various problems with Kubernetes:

1. UI not accessible

2. No nodes registered

    ```
    $ kubectl get nodes

    # no results
    ```

3. Errors running kubectl

    ```
    $ kubectl get po
    Error from server: an error on the server has prevented the request from succeeding
    ```
    
@BrianHicks @zogg any ideas on this?

**Comment by SillyMoo** [comment]
I get the same issue, but if I re-run with a 'vagrant provision' it all springs to life. Looks like a timing issue to me (I know that the ansible scripts wait for hyperkube to be pulled, but do they start for it to actually be up and listening?).

**Comment by SillyMoo** [comment]
Ok, I tell a bit of a lie. Ansible finishes ok, hyperkube is running and I see a node in kubectl. However I can't actually get kubernetes to pull any images (the pod just sits there, no image pull events, and no sign on the kubeworker that any images are being pulled).

**Comment by andreimc** [comment]
With latest master merged in it fails to restart skydns, not sure what would be causing it It just hangs for a while then I get the following error message:

```
NOTIFIED: [dnsmasq | restart dnsmasq] *****************************************
changed: [control-01]
changed: [kubeworker-001]
changed: [edge-001]

PLAY [role=worker] ************************************************************

TASK: [mesos | install mesos packages] ****************************************
FATAL: no hosts matched or all hosts have already failed -- aborting

```

Not sure why.

**Comment by stevendborrelli** [comment]
Docker fails on this 

```
TASK: [docker | enable docker] ************************************************ 
failed: [control-01] => {"failed": true}
msg: Job for docker.service failed because a configured resource limit was exceeded. See "systemctl status docker.service" and "journalctl -xe" for details.

```

But this is due to the new docker implementation not creating a `/etc/sysconfig/mantl-storage` file on non-lvm based systems and not a problem with this PR.

**Comment by andreimc** [comment]
The problems with this will be fixed after #1409 and #1410 are merged in.

**Comment by siddharthist** [comment]
@andreijs @andreimc Can you rebase this? Both those PRs have been merged.

**Comment by andreimc** [comment]
@siddharthist up to date.

**Comment by ryane** [comment]
Had a successful build but I am back to 

```
Internal Server Error (500)

Get https://10.254.0.1:443/api/v1/replicationcontrollers: dial tcp 10.254.0.1:443: getsockopt: connection refused
```

when trying to access the Kubernetes UI. 10.254.0.1 is the cluster ip for the `kubernetes` service. 

```
kubectl get svc --namespace=default
NAME         CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   10.254.0.1   <none>        443/TCP   1h
```

Logs from the `kubernetes-dashboard` pod's container:

```
2016/05/04 12:50:26 Incoming HTTP/1.0 GET /api/v1/replicationcontrollers request from 10.10.99.1:33787
2016/05/04 12:50:26 Getting list of all replication controllers in the cluster
2016/05/04 12:50:26 Get https://10.254.0.1:443/api/v1/replicationcontrollers: dial tcp 10.254.0.1:443: getsockopt: connection refused
2016/05/04 12:50:26 Outcoming response to 10.10.99.1:33787 with 500 status code
```

**Comment by andreimc** [comment]
hey guys, I had some time span this up in vagrant, I still get 502 for kube ui. :(

**Comment by andreimc** [comment]
I get the following on control-01 when i try to list pods:

[vagrant@control-01 ~]$ kubectl --namespace kube-system get pods
Error from server: an error on the server has prevented the request from succeeding

**Comment by manishrajkarnikar** [comment]
@andreimc @siddharthist curious how  is file groups_var/all/kubernetes_vars.yml read in vagrant run? or is it required at all.

**Comment by Zogg** [comment]
@manishrajkarnikar `groups_var/all/kubernetes_vars.yml` usage in local ansible run shouldn't be different from the remote case. 
As far as I remember yes, the kubernetes_vars.yml was mandatory to have the kubernetes roles play nice.

**Comment by manishrajkarnikar** [comment]
@Zogg I don't see it being mentioned in the vagrant file. I added that in my vagrant file as raw parameter and I was able to get K8s multi node cluster up and running. I couldn't get single master and slave node going though probably because of bug reported in vagrant file.

**Comment by andreijs** [comment]
Opening new PR of a branch closing this
